File: editorial/chap03/python/answer_B13.py
# A[l]에서 A[r]까지의 합계값
def sum(l, r, S):
	return S[r+1] - S[l]

# 입력
N, K = map(int, input().split())
A = list(map(int, input().split()))

# 누적 합을 구한다
S = [ 0 ] * (N + 1)
for i in range(1, N+1):
	S[i] = S[i-1] + A[i-1]

# 배열 준비(R은 0번부터 시작한다는 점에 유의)
R = [ None ] * N

# 자벌레 알고리즘
for i in range(N):
	if i == 0:
		R[i] = -1
	else:
		R[i] = R[i - 1]
	while R[i] < N-1 and sum(i,R[i]+1,S) <= K:
		R[i] += 1

# 답을 구한다
Answer = 0
for i in range(N):
	Answer += (R[i] - i + 1)
print(Answer)

File: editorial/chap03/cpp/answer_B13.cpp
#include <iostream>
using namespace std;
 
long long N, K;
long long A[100009];
long long S[100009]; // 누적 합계
long long R[100009]; // 왼쪽 끝이 정해지면 오른쪽 끝은 어디까지 갈 수 있을까
 
// A[l]에서 A[r]까지의 합계값
long long sum(int l, int r) {
	return S[r] - S[l - 1];
}
 
int main() {
	// 입력
	cin >> N >> K;
	for (int i = 1; i <= N; i++) cin >> A[i];
 
	// 누적 합을 구한다
	S[0] = 0;
	for (int i = 1; i <= N; i++) S[i] = S[i - 1] + A[i];
 
	// 자벌레 알고리즘
	for (int i = 1; i <= N; i++) {
		if (i == 1) R[i] = 0;
		else R[i] = R[i - 1];
		while (R[i] < N && sum(i, R[i] + 1) <= K) {
			R[i] += 1;
		}
	}
	
	// 답을 구한다
	long long Answer = 0;
	for (int i = 1; i <= N; i++) Answer += (R[i] - i + 1);
	cout << Answer << endl;
	return 0;
}
